internal/controlplane/handlers_oauth.go

```diff
@@ -221,34 +264,111 @@ func (s *Server) processCallback(ctx context.Context, w http.ResponseWriter, r *
 	return nil
 }
 
-func (s *Server) generateOAuthToken(ctx context.Context, providerName string, code string,
-	stateData db.GetProjectIDBySessionStateRow) error {
-	// generate a new OAuth2 config for the given provider
-	oauthConfig, err := s.providerAuthFactory(providerName, true)
+func (s *Server) processAppCallback(ctx context.Context, w http.ResponseWriter, r *http.Request,
+	pathParams map[string]string) error {
+	state := r.URL.Query().Get("state")
+	code := r.URL.Query().Get("code")
+
+	// Configure tracing
+	span := trace.SpanFromContext(ctx)
+	span.SetName("server.HandleGitHubAppCallback")
+	span.SetAttributes(attribute.Key("code").String(code))
+	defer span.End()
+
+	provider := pathParams["provider"]
+
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = provider
+
+	token, err := s.exchangeCodeForToken(ctx, provider, code)
 	if err != nil {
-		return fmt.Errorf("error creating OAuth config: %w", err)
-	}
-	if oauthConfig == nil {
-		return errors.New("oauth2.Config is nil")
+		return err
 	}
 
-	token, err := oauthConfig.Exchange(ctx, code)
+	installationID, err := strconv.ParseInt(r.URL.Query().Get("installation_id"), 10, 64)
 	if err != nil {
-		return fmt.Errorf("error exchanging code for token: %w", err)
+		return fmt.Errorf("unable to parse installation ID to integer: %v", err)
 	}
 
-	_, err = s.providers.CreateGitHubOAuthProvider(ctx, providerName, db.ProviderClassGithub, *token, stateData)
+	err = s.providers.ValidateGitHubInstallationId(ctx, token, installationID)
 	if err != nil {
-		if errors.Is(err, providers.ErrInvalidTokenIdentity) {
-			return newHttpError(http.StatusForbidden, "User token mismatch").SetContents(
-				"The provided login token was associated with a different GitHub user.")
+		return newHttpError(http.StatusForbidden, "User installation ID mismatch").SetContents(
+			"The GitHub user does not have access to the requested installation.")
+	}
+
+	if state != "" {
+		// Check the nonce to make sure it's valid
+		valid, err := mcrypto.IsNonceValid(state, s.cfg.Auth.NoncePeriod)
+		if err != nil {
+			return fmt.Errorf("error checking nonce: %w", err)
+		}
+		if !valid {
+			return errors.New("invalid nonce")
+		}
+
+		// get projectID from session along with state nonce from the database
+		stateData, err := s.store.GetProjectIDBySessionState(ctx, state)
+		if err != nil {
+			return fmt.Errorf("error getting project ID by session state: %w", err)
+		}
```

Comment:
This block seems like it could be a sub-block:

```golang
// returns non-nil error if state is not valid.  Free function, could live wherever.
stateData, err := getValidSessionState(ctx, state, s.cfg.Auth)
```